Sources/FoundationEssentials/FileManager/FileManager+Files.swift

```diff
@@ -276,6 +276,9 @@ extension _FileManagerImpl {
             }
             attr?[.protectionKey] = nil
         }
+        #elseif os(WASI)
+        // WASI doesn't support a temporary file, so can't use `.atomic`
+        let opts: Data.WritingOptions = []
```

Comment:
It seems unfortunate to allow a caller to think they are getting the safety of an atomic write and then silently drop it. Maybe it would be better to make the atomic option unavailable in WASI so that the caller has to knowingly not use it.

Reply:
@parkera Thank you for reviewing the PR. I'm sorry if I don't understanding what you are saying correctly, but do you mean something like this?

```diff
diff --git a/Sources/FoundationEssentials/Data/Data.swift b/Sources/FoundationEssentials/Data/Data.swift
index 70df1e8..55a4e33 100644
--- a/Sources/FoundationEssentials/Data/Data.swift
+++ b/Sources/FoundationEssentials/Data/Data.swift
@@ -2056,8 +2056,10 @@ public struct Data : Equatable, Hashable, RandomAccessCollection, MutableCollect
         public let rawValue: UInt
         public init(rawValue: UInt) { self.rawValue = rawValue }
 
+#if !os(WASI)
         /// An option to write data to an auxiliary file first and then replace the original file with the auxiliary file when the write completes.
         public static let atomic = WritingOptions(rawValue: 1 << 0)
+#endif
         
         /// An option that attempts to write data to a file and fails with an error if the destination file already exists.
         public static let withoutOverwriting = WritingOptions(rawValue: 1 << 1)
```

<details>
<summary>full diff</summary>

```diff
diff --git a/Benchmarks/Benchmarks/DataIO/BenchmarkDataIO.swift b/Benchmarks/Benchmarks/DataIO/BenchmarkDataIO.swift
index e660f8a..8bf2cd4 100644
--- a/Benchmarks/Benchmarks/DataIO/BenchmarkDataIO.swift
+++ b/Benchmarks/Benchmarks/DataIO/BenchmarkDataIO.swift
@@ -88,9 +88,11 @@ let benchmarks = {
         try data.write(to: testPath)
     }
     
+#if !os(WASI)
     Benchmark("write-regularFile-atomic", configuration: .cleanupTestPathConfig) { benchmark in
         try data.write(to: testPath, options: .atomic)
     }
+#endif
     
     Benchmark("write-regularFile-alreadyExists",
               configuration: .init(
@@ -103,6 +105,7 @@ let benchmarks = {
         try? data.write(to: testPath)
     }
     
+#if !os(WASI)
     Benchmark("write-regularFile-alreadyExists-atomic",
               configuration: .init(
                 setup: {
@@ -113,6 +116,7 @@ let benchmarks = {
     ) { benchmark in
         try? data.write(to: testPath, options: .atomic)
     }
+#endif
     
     Benchmark("read-regularFile", 
               configuration: .init(
diff --git a/Sources/FoundationEssentials/Data/Data+Writing.swift b/Sources/FoundationEssentials/Data/Data+Writing.swift
index 7c23e1e..ff2c4c9 100644
--- a/Sources/FoundationEssentials/Data/Data+Writing.swift
+++ b/Sources/FoundationEssentials/Data/Data+Writing.swift
@@ -314,13 +314,18 @@ internal func writeToFile(path inPath: PathOrURL, data: Data, options: Data.Writ
 }
 
 internal func writeToFile(path inPath: PathOrURL, buffer: UnsafeRawBufferPointer, options: Data.WritingOptions, attributes: [String : Data] = [:], reportProgress: Bool = false) throws {
+#if os(WASI)
+    try writeToFileNoAux(path: inPath, buffer: buffer, options: options, attributes: attributes, reportProgress: reportProgress)
+#else
     if options.contains(.atomic) {
         try writeToFileAux(path: inPath, buffer: buffer, options: options, attributes: attributes, reportProgress: reportProgress)
     } else {
         try writeToFileNoAux(path: inPath, buffer: buffer, options: options, attributes: attributes, reportProgress: reportProgress)
     }
+#endif
 }
 
+#if !os(WASI)
 /// Create a new file out of `Data` at a path, using atomic writing.
 private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPointer, options: Data.WritingOptions, attributes: [String : Data], reportProgress: Bool) throws {
     assert(options.contains(.atomic))
@@ -495,7 +500,6 @@ private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPoint
                 
                 cleanupTemporaryDirectory(at: temporaryDirectoryPath)
                 
-#if !os(WASI) // WASI does not support fchmod for now
                 if let mode {
                     // Try to change the mode if the path has not changed. Do our best, but don't report an error.
 #if FOUNDATION_FRAMEWORK
@@ -519,16 +523,18 @@ private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPoint
                     fchmod(fd, mode)
 #endif
                 }
-#endif // os(WASI)
             }
         }
     }
 #endif
 }
+#endif  // os(WASI)
 
 /// Create a new file out of `Data` at a path, not using atomic writing.
 private func writeToFileNoAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPointer, options: Data.WritingOptions, attributes: [String : Data], reportProgress: Bool) throws {
+#if !os(WASI)
     assert(!options.contains(.atomic))
+#endif
 
 #if os(Windows)
     try inPath.path.withNTPathRepresentation { pwszPath in
diff --git a/Sources/FoundationEssentials/Data/Data.swift b/Sources/FoundationEssentials/Data/Data.swift
index 70df1e8..55a4e33 100644
--- a/Sources/FoundationEssentials/Data/Data.swift
+++ b/Sources/FoundationEssentials/Data/Data.swift
@@ -2056,8 +2056,10 @@ public struct Data : Equatable, Hashable, RandomAccessCollection, MutableCollect
         public let rawValue: UInt
         public init(rawValue: UInt) { self.rawValue = rawValue }
 
+#if !os(WASI)
         /// An option to write data to an auxiliary file first and then replace the original file with the auxiliary file when the write completes.
         public static let atomic = WritingOptions(rawValue: 1 << 0)
+#endif
         
         /// An option that attempts to write data to a file and fails with an error if the destination file already exists.
         public static let withoutOverwriting = WritingOptions(rawValue: 1 << 1)
@@ -2442,9 +2444,11 @@ public struct Data : Equatable, Hashable, RandomAccessCollection, MutableCollect
     /// - parameter options: Options for writing the data. Default value is `[]`.
     /// - throws: An error in the Cocoa domain, if there is an error writing to the `URL`.
     public func write(to url: URL, options: Data.WritingOptions = []) throws {
+#if !os(WASI)
         if options.contains(.withoutOverwriting) && options.contains(.atomic) {
             fatalError("withoutOverwriting is not supported with atomic")
         }
+#endif
         
         guard url.isFileURL else {
             throw CocoaError(.fileWriteUnsupportedScheme)
diff --git a/Sources/FoundationEssentials/String/String+IO.swift b/Sources/FoundationEssentials/String/String+IO.swift
index 4adc11b..a210e14 100644
--- a/Sources/FoundationEssentials/String/String+IO.swift
+++ b/Sources/FoundationEssentials/String/String+IO.swift
@@ -447,7 +447,12 @@ extension StringProtocol {
             attributes = [:]
         }
 
+#if os(WASI)
+        guard !useAuxiliaryFile else { throw CocoaError(.featureUnsupported) }
+        let options : Data.WritingOptions = []
+#else
         let options : Data.WritingOptions = useAuxiliaryFile ? [.atomic] : []
+#endif
 
         try writeToFile(path: .path(String(path)), data: data, options: options, attributes: attributes, reportProgress: false)
     }
@@ -465,7 +470,12 @@ extension StringProtocol {
             attributes = [:]
         }
 
+#if os(WASI)
+        guard !useAuxiliaryFile else { throw CocoaError(.featureUnsupported) }
+        let options : Data.WritingOptions = []
+#else
         let options : Data.WritingOptions = useAuxiliaryFile ? [.atomic] : []
+#endif
 
         try writeToFile(path: .url(url), data: data, options: options, attributes: attributes, reportProgress: false)
     }
diff --git a/Tests/FoundationEssentialsTests/DataIOTests.swift b/Tests/FoundationEssentialsTests/DataIOTests.swift
index 999ecea..c515cd7 100644
--- a/Tests/FoundationEssentialsTests/DataIOTests.swift
+++ b/Tests/FoundationEssentialsTests/DataIOTests.swift
@@ -91,6 +91,7 @@ class DataIOTests : XCTestCase {
         cleanup(at: url)
     }
 
+#if !os(WASI)
     // Atomic writing is a very different code path
     func test_readWriteAtomic() throws {
         let url = testURL()
@@ -102,6 +103,7 @@ class DataIOTests : XCTestCase {
 
         cleanup(at: url)
     }
+#endif
 
     func test_readWriteMapped() throws {
         let url = testURL()
```

</details>

Reply:
Can we use availability annotations for WASI yet?
```
@available(wasi?, unavailable, message: "atomic writing is unavailable in WASI")
```

Reply:
> Can we use availability annotations for WASI yet?

No, we can't yet. So I use `#if !os(WASI)` instead.

https://github.com/swiftlang/swift/blob/80190ad5397c7cf3bfe07ac21d3aa550679e9afd/include/swift/AST/PlatformKinds.def#L25-L38

Reply:
Is there some kind of fundamental limitation there or can we just add to that list? Especially considering we support `os(WASI)` already.

Reply:
> I would also be curious to learn more about why we can't support the option on WASI in the first place. No mkstemp?

Yes, WASI specs don't include the concept of a temporary file and APIs like `mkstemp`.

https://github.com/WebAssembly/WASI/issues/306

It is not impossible to access `/tmp` on the host if the Wasm runtime maps it to the file system on the guest, but there seems to be a problem that it doesn't match the mechanism for sandboxing. Whether or not `/tmp` can be accessible depends on the runtime and its options, so there is no guarantee that it can always be accessible.

Discussions: https://github.com/swiftwasm/swift/issues/5593#issuecomment-2412799109

Reply:
> Is there some kind of fundamental limitation there or can we just add to that list? Especially considering we support os(WASI) already.

First, I think it is not enough to simply add WASI to the list, as WASI has some variations (e.g. wasip1, wasip2, wasip1-threads, ...). So I don't think it is that easy.

Second, `@available` is evaluated at runtime, so it must be distinguished from other WebAssembly environments (e.g. `wasm32-unknown-none-wasm`). In contrast, `os(WASI)` is evaluated at compile time, so it can be easily determined by a target triple (e.g. `wasm32-unknown-wasi`, `wasm32-unknown-wasip1-threads`).

Reply:
The difference is that if we define it away, the error is "that doesn't exist", vs "that is not available for your OS". The latter is a much more understandable message, and why we have availability annotations in the first place.

Reply:
@stephentyrone suggests this instead in the short term:

```
#if !os(WASI)
/// An option to write data to an auxiliary file first and then replace the original file with the auxiliary file when the write completes.
public static let atomic = WritingOptions(rawValue: 1 << 0)
#else
@available(*, unavailable, message: "atomic writing is unavailable in WASI")
public static let atomic = WritingOptions(rawValue: 1 << 0)
#endif
```

Reply:
Or this, if it compiles

```
#if os(WASI)
@available(*, unavailable, message: "atomic writing is unavailable in WASI")
#endif
/// An option to write data to an auxiliary file first and then replace the original file with the auxiliary file when the write completes.
public static let atomic = WritingOptions(rawValue: 1 << 0)
```

Reply:
Then we'll have to follow that through all the places we use the atomic option in Foundation itself. But at least it's honest.

Reply:
I agree with your idea itself. But it will take much more time and effort to make it happen, at least for me. So, if you decide that we need to add `@available(wasi)` first, I think it would be better to close this pull request at this time.

By the way, if I wanted to add `available(wasi)`, should I submit a proposal to the Swift evolution process?

Reply:
Let's try the workaround above for the short term. But yes, Swift needs to support availability annotations for WASI (and it can have versions, just like macOS or iOS does) - and that is a process we should start in parallel.

Reply:
Thank you for your suggestion! (and thank you @stephentyrone)

I'll try it.

Reply:
@parkera How about this commit? 77feb1c943b6e71c6435da9eb7fb9e73074398b1

I found that the following code causes an error. So the diff is a little bit larger.

```swift
@available(*, unavailable, message: "foo is unavailable")
func foo() {}

@available(*, unavailable, message: "bar is unavailable")
func bar() {
    foo()  // <-- error: 'foo()' is unavailable: foo is unavailable
}
```

I confirmed that `swift build` succeeded.

```console
$ swift --version
Swift version 6.0.3-dev (LLVM 81a2745a5c40b5f, Swift 8f95e04bb5d9d9a)
Target: aarch64-unknown-linux-gnu
$ swift sdk list
6.0-SNAPSHOT-2024-10-10-a-wasm32-unknown-wasi
$ swift build --swift-sdk wasm32-unknown-wasi
...
Build complete! (82.28s)
$ swift build
...
Build complete! (61.87s)
```